### PR TITLE
8279888: Local variable independently used by multiple loops can interfere with loop optimizations

### PR DESCRIPTION
**Pull request by rwestrel**
The bytecode of the 2 methods of the benchmark is structured
differently: loopsWithSharedLocal(), the slowest one, has multiple
backedges with a single head while loopsWithScopedLocal() has a single
backedge and all the paths in the loop body merge before the
backedge. loopsWithSharedLocal() has its head cloned which results in
a 2 loops loop nest.

loopsWithSharedLocal() is slow when 2 of the backedges are most
commonly taken with one taken only 3 times as often as the other
one. So a thread executing that code only runs the inner loop for a
few iterations before exiting it and executing the outer loop. I think
what happens is that any time the inner loop is entered, some
predicates are executed and the overhead of the setup of loop strip
mining (if it's enabled) has to be paid. Also, if iteration
splitting/unrolling was applied, the main loop is likely never
executed and all time is spent in the pre/post loops where potentially
some range checks remain.

The fix I propose is that ciTypeFlow, when it clone heads, not only
rewires the most frequent loop but also all this other frequent loops
that share the same head. loopsWithSharedLocal() and
loopsWithScopedLocal() are then fairly similar once c2 parses them.

Without the patch I measure:

LoopLocals.loopsWithScopedLocal      mixed  avgt    5  1108.874 ± 250.463  ns/op
LoopLocals.loopsWithSharedLocal      mixed  avgt    5  1575.665 ±  70.372  ns/op

with it:

LoopLocals.loopsWithScopedLocal      mixed  avgt    5  1108.180 ± 245.873  ns/op
LoopLocals.loopsWithSharedLocal      mixed  avgt    5  1234.665 ± 157.912  ns/op

But this patch also causes a regression when running one of the
benchmarks added by 8278518. From:

SharedLoopHeader.sharedHeader  avgt    5  505.993 ± 44.126  ns/op

to:

SharedLoopHeader.sharedHeader  avgt    5  724.253 ± 1.664  ns/op

The hot method of this benchmark used to be compiled with 2 loops, the
inner one a counted loop. With the patch, it's now compiled with a
single one which can't be converted into a counted loop because the
loop variable is incremented by a different amount along the 2 paths
in the loop body. What I propose to fix this is to add a new loop
transformation that detects that, because of a merge point, a loop
can't be turned into a counted loop and transforms it into 2
loops. The benchmark performs better with this:

SharedLoopHeader.sharedHeader  avgt    5  567.150 ± 6.120  ns/op

Not quite on par with the previous score but AFAICT this is due to
code generation not being as good (the loop head can't be aligned in
particular).

In short, I propose:

- changing ciTypeFlow so that, when it pays off, a loop with
multiple backedges is compiled as a single loop with a merge point in
the loop body

- adding a new loop transformation so that, when it pays off, a loop
with a merge point in the loop body is converted into a 2 loops loop
nest, essentially the opposite transformation.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279888](https://bugs.openjdk.java.net/browse/JDK-8279888): Local variable independently used by multiple loops can interfere with loop optimizations

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Contributors
 * Claes Redestad `<redestad@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7352/head:pull/7352` \
`$ git checkout pull/7352`

Update a local copy of the PR: \
`$ git checkout pull/7352` \
`$ git pull https://git.openjdk.java.net/jdk pull/7352/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7352`

View PR using the GUI difftool: \
`$ git pr show -t 7352`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7352.diff">https://git.openjdk.java.net/jdk/pull/7352.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
One thing I should have mentioned is that I refactored loop body cloning so that it now calls helper functions. The new loop transform is implemented by calling these functions so it's mostly based on well tested code.

**Comment by TobiHartmann** [comment]
I executed some testing and I'm seeing massive (> 700) failures due to the `negative trip count?` assert. For example, `compiler/c2/Test6603011.java` with `-server -Xmixed`:
```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (.../open/src/hotspot/share/opto/loopPredicate.cpp:1447), pid=11690, tid=11708
#  assert(!follow_branches || loop_trip_cnt >= 0) failed: negative trip count?
#
# JRE version: Java(TM) SE Runtime Environment (19.0) (fastdebug build 19-internal+0-2022-02-08-1039057.tobias.hartmann.jdk2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 19-internal+0-2022-02-08-1039057.tobias.hartmann.jdk2, compiled mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x135c470]  PhaseIdealLoop::loop_predication_impl(IdealLoopTree*) [clone .part.0]+0xf70

Current CompileTask:
C2:  10125 4309    b  4       jdk.internal.org.objectweb.asm.ClassReader::readUtf (161 bytes)

Stack: [0x00007f83f4c3a000,0x00007f83f4d3b000],  sp=0x00007f83f4d34c60,  free space=1003k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x135c470]  PhaseIdealLoop::loop_predication_impl(IdealLoopTree*) [clone .part.0]+0xf70
V  [libjvm.so+0x135c739]  IdealLoopTree::loop_predication(PhaseIdealLoop*)+0x109
V  [libjvm.so+0x13a4b47]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0x1327
V  [libjvm.so+0xa9cbea]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x28a
V  [libjvm.so+0xa99513]  Compile::Optimize()+0x1193
V  [libjvm.so+0xa9b5f5]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1575
V  [libjvm.so+0x8b1e29]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x669
V  [libjvm.so+0xaab4d8]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xc88
V  [libjvm.so+0xaac2b8]  CompileBroker::compiler_thread_loop()+0x668
V  [libjvm.so+0x19382da]  JavaThread::thread_main_inner()+0x25a
V  [libjvm.so+0x1940620]  Thread::call_run()+0x100
V  [libjvm.so+0x1623cd4]  thread_native_entry(Thread*)+0x104
```

**Comment by rwestrel** [comment]
> I executed some testing and I'm seeing massive (> 700) failures due to the `negative trip count?` assert. For example, `compiler/c2/Test6603011.java` with `-server -Xmixed`:

Thanks for running tests. A last minute tweak that's not as harmless as I thought. It should be fixed now.

**Comment by TobiHartmann** [comment]
Testing now passed but I need more time to review this.

**Comment by rwestrel** [comment]
Anyone for a review?

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -2494,68 +2529,67 @@ void PhaseIdealLoop::clone_loop( IdealLoopTree *loop, Node_List &old_new, int dd
...
 
         // Now finish up 'r'
-        r->set_req( 1, newuse );
-        r->set_req( 2,    use );
+        r->set_req( 1, newuse);
+        r->set_req( 2,    use);
         _igvn.register_new_node_with_optimizer(r);
         set_loop(r, use_loop);
-        set_idom(r, !side_by_side_idom ? newuse->in(0) : side_by_side_idom, dd_r);
+        set_idom(r, side_by_side_idom == NULL? newuse->in(0) : side_by_side_idom, dd_r);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2536]:
```suggestion
        set_idom(r, (side_by_side_idom == NULL) ? newuse->in(0) : side_by_side_idom, dd_r);
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -2494,68 +2529,67 @@ void PhaseIdealLoop::clone_loop( IdealLoopTree *loop, Node_List &old_new, int dd
         }
 
         // Now finish up 'r'
-        r->set_req( 1, newuse );
-        r->set_req( 2,    use );
+        r->set_req( 1, newuse);
+        r->set_req( 2,    use);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2533]:
```suggestion
        r->set_req(1, newuse);
        r->set_req(2,    use);
```

**Review by TobiHartmann — COMMENTED** [review]
Looks good to me overall but this is very hard to review. More people should look at it.

If you resolve the merge conflicts, I can run some more testing with `-XX:+StressDuplicateBackedge` enabled.

**Comment by rwestrel** [comment]
> Looks good to me overall but this is very hard to review. More people should look at it.
> 
> If you resolve the merge conflicts, I can run some more testing with `-XX:+StressDuplicateBackedge` enabled.

Thanks for reviewing this. I applied your suggestions and merged.

**Comment by TobiHartmann** [comment]
`compiler/intrinsics/unsafe/HeapByteBufferTest.java` fails with `-XX:+StressDuplicateBackedge`:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (...workspace/open/src/hotspot/share/opto/macroArrayCopy.cpp:1277), pid=23931, tid=23946
#  assert(alloc != __null) failed: expect alloc
#
# JRE version: Java(TM) SE Runtime Environment (19.0) (fastdebug build 19-internal-2022-03-23-1109324.tobias.hartmann.jdk2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 19-internal-2022-03-23-1109324.tobias.hartmann.jdk2, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x13f86f5]  PhaseMacroExpand::expand_arraycopy_node(ArrayCopyNode*)+0x10a
```

**Comment by rwestrel** [comment]
> compiler/intrinsics/unsafe/HeapByteBufferTest.java

Thanks for running tests. That one should be fixed now.

**Comment by TobiHartmann** [comment]
Thanks, I'll re-run testing.

**Comment by TobiHartmann** [comment]
All tests passed.

File: src/hotspot/share/opto/c2_globals.hpp [hunk 3]
```diff
@@ -765,6 +765,14 @@
           "for at most jint_max / StressLongCountedLoop")                   \
           range(0, max_juint)                                               \
                                                                             \
+  product(bool, DuplicateBackedge, true,                                    \
```
Comment (vnkozlov) [inline src/hotspot/share/opto/c2_globals.hpp:768]: Why flag is `product`?  Can it be `experimental` or `diagnostic`? I assume eventually we should remove this flag.

**Review by vnkozlov — COMMENTED** [review]
Nice work. I have only comment about the flag.

File: src/hotspot/share/opto/c2_globals.hpp (hunk 3 above)
Comment (rwestrel) [reply src/hotspot/share/opto/c2_globals.hpp]: Right. I made it diagnostic in the new commit.

**Comment by rwestrel** [comment]
> Nice work. I have only comment about the flag.

Thanks for reviewing this @vnkozlov

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TobiHartmann** [comment]
Looks reasonable to me. I'll re-run some testing and report back.

**Comment by TobiHartmann** [comment]
Tests all passed.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**